portfolio/src/main/java/com/google/sps/servlets/FileHandlerServlet.java

```diff
@@ -0,0 +1,106 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import com.google.appengine.repackaged.com.google.gson.Gson;
+import com.google.appengine.api.datastore.Query;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/files")
+public class FileHandlerServlet extends HttpServlet {
+
+private static final String CONTENT_TYPE_JSON = "application/json;";
+private static final String KEY_IMAGE_BLOBKEY = "blobKey";
+private static final String KEY_IMAGE_URL = "url";
+private static final String KIND_IMAGE = "image";
+private static final String PATH_HOME_PAGE = "/index.html";
+private final BlobstoreService blobService = BlobstoreServiceFactory.getBlobstoreService();
+private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+@Override
+protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+  List<Entity> files = datastore.prepare(new Query(KIND_IMAGE)).asList(FetchOptions.Builder.withDefaults());
+  String[] urls = files
+                    .stream()
+                    .map(e -> e.getProperty(KEY_IMAGE_URL))
+                    .toArray(String[]::new);
+  resp.setContentType(CONTENT_TYPE_JSON);
+  resp.getWriter().println(new Gson().toJson(urls));
+}
+
+ @Override
+ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+     Map<String, List<BlobKey>> blobs = blobService.getUploads(req);
+     List<BlobKey> blobKeys = blobs.get(KIND_IMAGE);
+
+     if (blobKeys == null || blobKeys.isEmpty()) {
+         // the user didn't upload a file, do nothing
+         resp.sendRedirect(PATH_HOME_PAGE);
+         return;
+     }
+
+     // The form only contains a single file input
+     BlobKey blobKey = blobKeys.get(0);
```

Comment:
what if there were multiple files? you might only care about 1, which is fine, but should you cleanup the rest from the blobstore anyways?

Reply:
it might be helpful to think of this in terms of security as well - without this you've given a malicious user the ability to essentially charge you for unlimited storage...

Reply:
Fixed